### PR TITLE
Change TCODLine implementation

### PR DESCRIPTION
**Pull request by sensualcoder**
Updates TCODLine to be more idiomatic C++ style and replaces deprecated C function calls

### CODE REVIEW

**Comment by HexDecimal** [comment]
I can't accept a PR that breaks ABI compatibility, but I'm open to a full rewrite of line handling as long the previous functions stay as they are.

I request that new C++ code is added as fully inline code in the `.h` headers.  Which is how the new C++ API is handled.

**Comment by HexDecimal** [comment]
Honestly, I think every method of line drawing currently in libtcod has problems in both the C and C++ API.

**Comment by sensualcoder** [comment]
Fair enough. I'll revert it to the old static functions and add a note to deprecate them, then add the updated functions.

I'll have a look around at the other line drawing code as well!

**Comment by sensualcoder** [comment]
Woops, forgot to add documentation to the new functions.

**Comment by HexDecimal** [comment]
I should explain better, or maybe I should make an issue to organize things.

Since line drawing is a pretty low-level task most of the code should be moved inline, even in C using C99 inline functions.  I haven't gotten used to how to implement these correctly.

For the callback-based functions.  The C callbacks need an additional userdata parameter to be useful.  C++ should support lambdas as callbacks.

C should have a function that outputs a full array of line indexes to a buffer.  C++ might use this function to return a std::vector of line coordinates.

C++ could have a dynamic iterator supporting ranged-based for loops to iterate over line coordinates without a heap buffer.

I'm not asking you to implement all of these, I'm just collecting my thoughts.

**Comment by HexDecimal** [comment]
I recommend leaving `bresenham.cpp` and `bresenham.hpp` untouched.  New C++ code should be added to the end of `bresenham.h` in a block like this:

```cpp
#ifdef __cplusplus
namespace tcod {
// [new C++ API]
}  // namespace tcod
#endif  // __cplusplus
```
You should also leave the samples alone for now until the new API is decided on.

**Comment by sensualcoder** [comment]
Is this more like what you were thinking?

**Comment by HexDecimal** [comment]
> Is this more like what you were thinking?

Yes, you've understood me well so far.

> Thanks for the responses by the way, this is my first time ever contributing to anything open source.

You're doing pretty good.  This isn't an easy repository to contribute source changes to.

File: src/libtcod/bresenham.h [hunk 1]
```diff
@@ -80,7 +80,67 @@ TCODLIB_API bool TCOD_line_step_mt(int* __restrict xCur, int* __restrict yCur, T
...
+#include <functional>
+#include <vector>
+
+namespace tcod {
+/**
+tcod::Line class
+
+Implements a lightweight bresenham line drawing algorithm.
+ */
+class TCODLIB_API Line {
```
Comment (HexDecimal) [inline src/libtcod/bresenham.h:96]: Just `Line` might be to generic.  I might call this class `Bresenham` or `BresenhamLine` instead.

File: src/libtcod/bresenham.h [hunk 2]
```diff
@@ -80,7 +80,67 @@ TCODLIB_API bool TCOD_line_step_mt(int* __restrict xCur, int* __restrict yCur, T
...
+
+  /**
+    tcod::Line callback function
+
+    \rst
+    .. cpp:function:: bool tcod::Line::line(int xFrom, int yFrom, int xTo, int yTo, const std::function<bool(int, int)>&
+    callback);
+    \endrst
+  */
+  static bool line(int xFrom, int yFrom, int xTo, int yTo, const std::function<bool(int, int)>& callback) {
```
Comment (HexDecimal) [inline src/libtcod/bresenham.h:127]: This static method should be it's own separate function instead of inside the class.

**Review by HexDecimal — CHANGES_REQUESTED** [review]

File: src/libtcod/bresenham.h (hunk 2 above)
Comment (HexDecimal) [reply src/libtcod/bresenham.h:127]: And again it would need a longer name mentioning `bresenham`.

**Comment by HexDecimal** [comment]
The remaining issues would be the `#include` placement and the documentation (the functions need to be referenced from the docs directory, but that area is especially a mess.)  I feel like it might go smoother if I handled those instead of having you do them.

**Comment by sensualcoder** [comment]
Should I put the cpp includes inside the first `#ifdef __cplusplus` block?

**Comment by HexDecimal** [comment]
> Should I put the cpp includes inside the first `#ifdef __cplusplus` block, on line 37?

I usually put them in their own block, since in some cases you'd put regular C includes below them.

**Comment by HexDecimal** [comment]
Anyway, if you don't want me to suddenly merge this then you should convert it into a draft.

**Review by HexDecimal — APPROVED** [review]

**Comment by HexDecimal** [comment]
I've merged the changes although I'm still indecisive on what to do with them.  They'll at least be available provisionally for the moment.

I'm curious what setup you have to develop libtcod on.  Do you have any current projects using libtcod's C++ API?

**Comment by sensualcoder** [comment]
I actually have been working on going through and updating the old C++
libtcod tutorial over on gitlab (
https://gitlab.com/libtcod-tutorials/libtcod-tutorials) while following the
newer Python tutorial. Which is what led me to update these in particular
since I had gotten to the part of the tutorial that uses them for map
creation.

**Comment by HexDecimal** [comment]
Honestly the C++ API still needs a lot of work.  You seem to be better off than most others if you're willing to modify libtcod itself since the current API is not going to lead to clean C++ code.

I don't have any C++ projects of my own yet since I mostly prefer writing in Python.  It took a lot of work just to set up libtcod with a decent C/C++ package manager.  I've been trying to prepare [a template project](https://github.com/HexDecimal/libtcod-vcpkg-template) for this.

The new Python tutorial was mostly designed around Python.  I'm not sure how easy it will be to translate it back into C++.

**Comment by HexDecimal** [comment]
I've raised issue #92 with more specifics on how I wanted line handling.  To me idiomatic C++ looks like this:
```cpp
for (auto&& [x, y] : tcod::bresenham({0, 0}, {4, 4})) {
  // ...
}
```

**Comment by sensualcoder** [comment]
I definitely want to help update the C++ API since that's my language of choice, so I'm happy to start digging in wherever.

I've actually found it rather easy to look at the Python tutorial code and translate it some to C++, much more so than trying to update from the older C++ tutorial on roguebasin.

**Pull request merged**